tests/core/test_encode_structured_data_vs_signTypedData_v4.py

```diff
@@ -0,0 +1,503 @@
+                "match": "The Primary Type `Person` is not present in the `types` attribute",  # noqa: E501
+            },
+            "message_to_encode": {
+                "types": {
+                    "EIP712Domain": [
+                        {"name": "name", "type": "string"},
+                    ],
+                    "Human": [
+                        {"name": "name", "type": "string"},
+                    ],
+                },
+                "primaryType": "Person",
+                "domain": {
+                    "name": "Name",
+                },
+                "message": {
+                    "name": "Bob",
+                },
+            },
+        },
+        {
+            "input_type": "Unrecognized non-primary type",
+            "expected_esd_hash": "2f1c830ad734e08c06419c229b6bb26e0fb4134b5c04a4dea2163171b79683eb",  # noqa: E501
+            "expected_esd_err": None,
+            "message_to_encode": {
+                "types": {
+                    "EIP712Domain": [
+                        {"name": "name", "type": "string"},
+                    ],
+                    "Person": [
+                        {"name": "name", "type": "string"},
+                    ],
+                },
+                "primaryType": "Person",
+                "domain": {
+                    "name": "Name",
+                },
+                "message": {
+                    "name": "Bob",
+                    "pet": {
+                        "animal": "cat",
+                        "age": 3,
+                    },
+                },
+            },
+        },
+        {
+            "input_type": "Extra type specified that isn't used by primary type",
+            "expected_esd_hash": "2f1c830ad734e08c06419c229b6bb26e0fb4134b5c04a4dea2163171b79683eb",  # noqa: E501
+            "expected_esd_err": None,
+            "message_to_encode": {
+                "types": {
+                    "EIP712Domain": [
+                        {"name": "name", "type": "string"},
+                    ],
+                    "Person": [
+                        {"name": "name", "type": "string"},
+                    ],
+                    "Pet": [
+                        {"name": "animal", "type": "string"},
+                        {"name": "age", "type": "uint256"},
+                    ],
+                },
+                "primaryType": "Person",
+                "domain": {
+                    "name": "Name",
+                },
+                "message": {
+                    "name": "Bob",
+                },
+            },
+        },
+    ),
+)
+def test_encode_structured_data(test_case):
+    (
+        input_type,
+        expected_esd_hash,
+        expected_esd_err,
+        message_to_encode,
+    ) = test_case.values()
+
+    if expected_esd_hash:
+        assert (
+            _hash_eip191_message(encode_structured_data(message_to_encode)).hex()
+            == expected_esd_hash
+        )
+    else:
```

Comment:
nit: We're moving away from having success and error cases in the same test, just for readability's sake. What do you think of splitting this into two tests, one for success cases, and one for errors?

Reply:
split!